pkg/cli/cli.go

```diff
@@ -1364,95 +1338,160 @@ func runConnectionWebSocketAgentCommand(flags *ConnectFlags, c *command, args []
 		c.PrintErrf("Failed to listen ADB socket: %v", err)
 		return err
 	}
+	device := args[0]
 	adbPort := l.Addr().(*net.TCPAddr).Port
-	adbErrCh := make(chan error)
-	go func(adbErrCh chan error) {
-		err := opts.ADBServerProxy.Connect(adbPort)
-		if err != nil {
-			c.PrintErrf("Failed to connect ADB to device: %v\n", err)
+	err = opts.ADBServerProxy.Connect(adbPort)
+	if err != nil {
+		c.PrintErrf("Failed to connect ADB to device: %v\n", err)
+	}
+
+	for {
+		if err := connectAndForwardOnce(flags, c, opts, l, adbPort, device); err != nil {
+			return err
+		}
+	}
+}
+
+func connectAndForwardOnce(
+	flags *ConnectFlags, c *command, opts *subCommandOpts, l net.Listener, adbPort int, device string) error {
+	// Get connection from ADB server
+	tcpConn, tcpErr := acceptADBConnection(opts, c, l, adbPort)
+	if tcpErr != nil {
+		return tcpErr
+	}
+	defer tcpConn.Close()
+
+	const maxRetryCount = 10
+	var retryCount int
+	var devConn io.ReadWriteCloser
+	var devErr error
+
+	// Try connect to remote device's ADB
+	for retryCount = 0; retryCount < maxRetryCount; retryCount++ {
+		devConn, devErr = connectToDeviceADB(flags, c, opts, device)
+		if devErr == nil {
+			break
 		}
-		adbErrCh <- err
-	}(adbErrCh)
+		c.PrintErrf("Failed to connect ADB to device: %v, retrying (%d / %d)\n", devErr, retryCount+1, maxRetryCount)
+		time.Sleep(time.Duration(math.Pow(2, float64(retryCount/2))) * time.Second)
+	}
+	if retryCount == maxRetryCount {
+		return devErr
+	}
+
+	if sendResultErr := sendConnectResult(c, adbPort); sendResultErr != nil {
+		devConn.Close()
+		return sendResultErr
+	}
+
+	forwardCh := make(chan interface{})
+	go func(forwardCh chan interface{}) {
+		io.Copy(devConn, tcpConn)
+		devConn.Close()
+		forwardCh <- nil
+	}(forwardCh)
+	io.Copy(tcpConn, devConn)
+	tcpConn.Close()
```

Comment:
not needed with the `defer` on line 1362.

Reply:
I used tcpConn.Close to interrupt io.Copy(...) in the goroutine above because in the main thread, it waits forwardCh and tio.Copy(...) function in the goroutine needs to be returned for avoiding blocking.

I changed `tcpConn.Close()` to `tcpConn.SetDeadline(time.Now())` to interrupt `io.Copy...` and avoid double closure.

Reply:
Good point, didn't notice that deadlock.

Could you leave a comment above the `SetDeadline` call explaining that it's intended to unblock the goroutine?

Reply:
Added comment about that.